tplink/hs300.go

```diff
@@ -141,3 +127,74 @@ func (p *HS300Strip) Pins(cap hal.Capability) ([]hal.Pin, error) {
 		return nil, fmt.Errorf("unsupported capability:%s", cap.String())
 	}
 }
+
+type hs300Factory struct {
+	meta       hal.Metadata
+	parameters []hal.ConfigParameter
+}
+
+var factory300 *hs300Factory
+var hs300once sync.Once
+
+// HS300Factory returns a singleton HS300 Driver factory
+func HS300Factory() hal.DriverFactory {
+
+	hs300once.Do(func() {
+		factory300 = &hs300Factory{
+			meta: hal.Metadata{
+				Name:        "tplink-hs300",
+				Description: "tplink hs300 series smart power strip driver with current monitoring",
+				Capabilities: []hal.Capability{
+					hal.DigitalOutput, hal.AnalogInput,
+				},
+			},
+			parameters: []hal.ConfigParameter{
+				{
+					Name:    "Address",
+					Type:    hal.String,
+					Order:   0,
+					Default: "192.168.1.11:9999",
+				},
+			},
+		}
+	})
+
+	return factory300
+}
+
+func (f *hs300Factory) Metadata() hal.Metadata {
+	return f.meta
+}
+
+func (f *hs300Factory) GetParameters() []hal.ConfigParameter {
+	return f.parameters
+}
+
+func (f *hs300Factory) ValidateParameters(parameters map[string]interface{}) (bool, map[string][]string) {
+
+	var failures = make(map[string][]string)
+
+	if v, ok := parameters["Address"]; ok {
+		_, ok := v.(string)
+		if !ok {
+			failure := fmt.Sprint("Address is not a string. ", v, " was received.")
+			failures["Address"] = append(failures["Address"], failure)
+		}
+	} else {
+		failure := fmt.Sprint("Address is a required parameter, but was not received.")
+		failures["Address"] = append(failures["Address"], failure)
+	}
+
+	return len(failures) == 0, failures
+}
+
+func (f *hs300Factory) NewDriver(parameters map[string]interface{}, hardwareResources interface{}) (hal.Driver, error) {
+	if valid, failures := f.ValidateParameters(parameters); !valid {
+		return nil, errors.New(hal.ToErrorString(failures))
+	}
+
+	addr := parameters["Address"].(string)
```

Comment:
check for typecasting error is not required?  If that's  a possibility then this will panic. Given we can error out, we can also check for  addr, ok := parameters[_address].(string)  and error out on !ok

Reply:
Typecast error should not be possible here since the parameters were validated just above.